### PR TITLE
New location detection 

### LINKED ISSUE #862: Node autodetects IP attributes of Proposal
- [ ] Mock IP attributes to Proposal and DiscoveryService
- [ ] Map location attributes from Location provider -> to Proposal same (way as "country" is mapped)

Location provider should return:
```
GET /location/me
{
  "ip": "1.2.3.4",
  "continent": "EU",
  "country": "LT",
  "city": "Vilnius",
  "asn": "62179",
  "user_type": "residential",
  "isp": "Telia Lietuva, AB"
}
```

Proposal should map those attributes to announced proposal structure:
```
// swagger:model ServiceLocationDTO
type locationRes struct {
	// Autonomous System Number
	// example: AS00001
	ASN string `json:"asn"`

	// example: NL
	Country string `json:"country,omitempty"`

	// example: Amsterdam
	City string `json:"city,omitempty"`

        // TODO add new location attributes
        Continent
        Country
        NodeType
        ISP
}
```

### LINKED ISSUE #863: Node integration with LocationOracle
- [x] Mock Tequilapi endpoint `/location` and `/connection/location`
- [x] Refactor CLI arg to have typed location providers in "--location.type"
- [x] Autodetect IP attributes from LocationOracle
- [x] Return IP attribytes in Tequilapi endpoints `/location` and `/connection/location`
- [x] Manual IP attributes configuration (can be done later)
- [x] Refactor CLI arg to have typed IP adapters in "--ip.type" (can be done later)
- [x] Stop using Ipify and migrate to LocationOracle (can be done later)

Node has several location detection providers:
```
--location.type         Location autodetect provider (builtin, mmdb, oracle) default: 'builtin'
--location.address   Address of specific location provider given in '--location.type' e.g. https://testnet-location.mysterium.network/location for --location.type=oracle provider

# Possibility from local binary database of GeoLite2 format e.g. http://dev.maxmind.com/geoip/geoip2/geolite2/
# would suggest localdb, we could eventually modify maxmind db with our own data.
--location.type=localdb

# Possibility to identity IP from Oracle
--location.type=oracle
--location.address="https://testnet-location.mysterium.network/location"

# Possible to have manual provider also:
--location.type=manual
--location.address=""
--location.country="LT"
--location.city="Vilnius"
--location.user_type="datacenter"
```

Node has several IP detection providers:
```
--ip.type         IP autodetect provider (ipify, oracle) default: 'ipify'
--ip.address   Address of specific IP provider given in --ip.type e.g. https://testnet-location.mysterium.network/ip for --ip.type=oracle provider e.g. https://api.ipify.org/ will be used for --ip.type=ipify

# Possibility to identity IP from Oracle
--ip.type=ipify
--ip.address="https://api.ipify.org/"


# Possibility to identity IP from Oracle
--ip.type=oracle
--ip.address="https://testnet-location.mysterium.network/ip"
```

### PR DESCRIPTION
**Pull request by soffokl**
Closes #863 
Closes #862

### CODE REVIEW

File: cmd/di.go [hunk 1]
```diff
@@ -460,20 +467,31 @@ func (di *Dependencies) bootstrapIdentityComponents(options node.Options) {
...
 	di.IPResolver = ip.NewResolver(options.IpifyUrl)
 
-	switch {
-	case options.Country != "":
-		di.LocationResolver = location.NewStaticResolver(options.Country)
-	case options.ExternalDb != "":
-		di.LocationResolver = location.NewExternalDbResolver(filepath.Join(configDirectory, options.ExternalDb))
+	switch options.Type {
+	case "localdb":
+		if options.ExternalDB != "" {
```
Comment (Waldz) [inline cmd/di.go]: I would like to have same `--location.address` for this

File: cmd/di.go [hunk 2]
```diff
@@ -460,20 +467,31 @@ func (di *Dependencies) bootstrapIdentityComponents(options node.Options) {
...
+func (di *Dependencies) bootstrapLocationComponents(options node.OptionsLocation, configDirectory string) error {
 	di.IPResolver = ip.NewResolver(options.IpifyUrl)
 
-	switch {
-	case options.Country != "":
-		di.LocationResolver = location.NewStaticResolver(options.Country)
-	case options.ExternalDb != "":
-		di.LocationResolver = location.NewExternalDbResolver(filepath.Join(configDirectory, options.ExternalDb))
+	switch options.Type {
+	case "localdb":
```
Comment (Waldz) [inline cmd/di.go]: Location autodetect provider (builtin, mmdb, oracle) default: 'builtin'

File: cmd/flags_location.go [hunk 3]
```diff
@@ -33,24 +33,50 @@ var (
...
 		Usage: "Service location country. If not given country is autodetected",
 		Value: "",
 	}
+	locationCityFlag = cli.StringFlag{
+		Name:  "location.city",
+		Usage: "Service location city",
+		Value: "",
+	}
+	locationNodeTypeFlag = cli.StringFlag{
+		Name:  "location.node.type",
```
Comment (Waldz) [inline cmd/flags_location.go]: `location.node-type`

File: core/location/oracle.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
```
Comment (Waldz) [inline core/location/oracle.go]: `core/location/oracle.go` -> `core/location/oracle_resolver.go`

File: tequilapi/endpoints/location.go [hunk 5]
```diff
@@ -28,8 +28,29 @@ import (
...
+	// Node Country
+	// example: LT
+	Country string
+	// Node City
+	// example: Vilnius
+	City string
+
+	// Node type
+	// example: residential
+	NodeType string
```
Comment (Waldz) [inline tequilapi/endpoints/location.go]: I suggest to call it `userType` and just in proposal `nodeType`

File: cmd/flags_location.go [hunk 6]
```diff
@@ -23,34 +23,54 @@ import (
...
-	// LocationCountryFlag allows to configure service country manually
-	LocationCountryFlag = cli.StringFlag{
+	locationAddressFlag = cli.StringFlag{
+		Name:  "location.address",
+		Usage: "Address of the service location system",
+		Value: "https://testnet-location.mysterium.network/api/v1/location",
+	}
+	locationCountryFlag = cli.StringFlag{
 		Name:  "location.country",
 		Usage: "Service location country. If not given country is autodetected",
```
Comment (soffokl) [inline cmd/flags_location.go]:
```suggestion
		Usage: "Service location country.",
```

File: core/location/oracle_resolver.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+type oracleResolver struct {
+	http                  mysterium.HTTPTransport
+	oracleResolverAddress string
+}
+
+// NewOracleResolver returns new db resolver initialized from Location Oracle service
+func NewOracleResolver(address string) *oracleResolver {
+	return &oracleResolver{
+		newHTTPTransport(1 * time.Minute),
```
[the remark this answers is not in the document]
Comment (soffokl) [reply core/location/oracle_resolver.go]: We have 1 minute for most of the REST calls, and some default value for HTTP 120 seconds. 
Looks like 1 minute is a good value.

File: core/location/db_resolver.go [hunk 8]
```diff
@@ -18,48 +18,54 @@
...
 }
 
-// ResolveCountry maps given ip to country
-func (r *DbResolver) ResolveCountry(ip string) (string, error) {
-	ipObject := net.ParseIP(ip)
-	if ipObject == nil {
-		return "", errors.New("failed to parse IP")
+// DetectLocation maps given ip to country
+func (r *DBResolver) DetectLocation(ip net.IP) (Location, error) {
+	if ip == nil {
```
[the remark this answers is not in the document]
Comment (soffokl) [reply core/location/db_resolver.go]: Changed it as described.

File: core/location/oracle_resolver.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		log.Error(oracleResolverLogPrefix, err)
+		return Location{}, err
+	}
+
+	err = o.doRequest(request, &location)
+	return location, err
+}
+
+func (o *oracleResolver) doRequest(request *http.Request, responseDto interface{}) error {
```
[the remark this answers is not in the document]
Comment (soffokl) [reply core/location/oracle_resolver.go]: Refactored it too.

**Pull request merged**